File: app/src/main/java/com/example/files/actions/DialogSort.java
package com.example.files.actions;

import static com.example.files.MainActivity.editor;
import static com.example.files.MainActivity.instance;
import static com.example.files.Statics.order;
import static com.example.files.Statics.sort;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.res.Configuration;
import android.icu.text.Collator;
import android.os.Handler;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Gravity;
import android.view.WindowManager;
import android.widget.TextView;

import com.example.files.JFileAdapter;
import com.example.files.R;
import com.example.files.models.JFile;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.button.MaterialButtonToggleGroup;

import java.util.ArrayList;
import java.util.Locale;

public class DialogSort extends BottomSheetDialog {

    Activity activity;
    ArrayList<JFile> jFileList;
    JFileAdapter jFileAdapter; // Recommended to sort the array before adding it to the adapter to prevent interrupting
    static Collator collator = Collator.getInstance(Locale.getDefault());;
    boolean isNormalSize;

    int checkedOrderButton = order, checkedSortButton = sort;

    public DialogSort() {
        super(instance);
        this.activity = instance;
        this.isNormalSize = isNormalScreen(activity);
    }

    public DialogSort(JFileAdapter jFileAdapter) {
        super(instance);
        this.activity = instance;
        this.jFileAdapter = jFileAdapter;
        this.jFileList = jFileAdapter.jFileList;
        this.isNormalSize = isNormalScreen(activity);
        createDialogSort();
    }

    @SuppressLint({"UseCompatLoadingForDrawables", "NonConstantResourceId"})
    private void createDialogSort() {

        setContentView(R.layout.dialog_sort);
//        setCancelable(true);
//        getWindow().setWindowAnimations(R.style.DialogAnimation);
//        getWindow().setBackgroundDrawable(activity.getDrawable(R.color.transparent));
//        getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT, WindowManager.LayoutParams.WRAP_CONTENT);
//        if (!activity.getResources().getConfiguration()
//                .isLayoutSizeAtLeast(Configuration.SCREENLAYOUT_SIZE_LARGE))
//            getWindow().setGravity(Gravity.BOTTOM);

        TextView apply = findViewById(R.id.sort_apply);
        TextView cancel = findViewById(R.id.sort_cancel);
        apply.setClipToOutline(true);
        cancel.setClipToOutline(true);

//        RadioGroup groupSort = findViewById(R.id.sort);
//        RadioGroup.LayoutParams layoutParams = new RadioGroup.LayoutParams(
//                RadioGroup.LayoutParams.MATCH_PARENT,
//                RadioGroup.LayoutParams.WRAP_CONTENT);

        MaterialButtonToggleGroup groupSort = findViewById(R.id.sort_selection);
        switch (sort) {
            case 0:
                groupSort.check(R.id.sort_by_name);
                break;
            case 1:
                groupSort.check(R.id.sort_by_size);
                break;
            case 2:
                groupSort.check(R.id.sort_by_date);
                break;
            case 3:
                groupSort.check(R.id.sort_by_type);
                break;
        }

        groupSort.addOnButtonCheckedListener((group, checkedId, isChecked) -> {
            if (isChecked) {
                switch (checkedId) {
                    case R.id.sort_by_name:
                        checkedSortButton = 0;
                        Log.d("##### checkedId #####", "sort_by_name");
                        break;
                    case R.id.sort_by_size:
                        checkedSortButton = 1;
                        Log.d("##### checkedId #####", "sort_by_size");
                        break;
                    case R.id.sort_by_date:
                        checkedSortButton = 2;
                        Log.d("##### checkedId #####", "sort_by_date");
                        break;
                    case R.id.sort_by_type:
                        checkedSortButton = 3;
                        Log.d("##### checkedId #####", "sort_by_type");
                        break;
                    default:
                        checkedSortButton = 0;
                        Log.d("##### checkedId #####", "default");
                }
            }
        });

        // add 5 radio buttons to the group
//        RadioButton radioButton;
//        for (int i = 0; i < 4; i++){
//            radioButton = new RadioButton(activity);
//            radioButton.setText(activity.getResources().getStringArray(R.array.sort)[i]);
//            radioButton.setId(i);
//            radioButton.setTextSize(16);
////            if (//!activity.getResources().getConfiguration().isLayoutSizeAtLeast(Configuration.SCREENLAYOUT_SIZE_LARGE) &&
////                    activity.getResources().getConfiguration().isLayoutSizeAtLeast(Configuration.SCREENLAYOUT_SIZE_NORMAL))
//            if (isNormalSize) radioButton.setPadding(0,40,30,40);
//            if (i == sort) radioButton.setChecked(true);
//            groupSort.addView(radioButton, layoutParams);
//        }

        MaterialButtonToggleGroup groupOrder = findViewById(R.id.order_selection);
        groupOrder.check(order == 0 ? R.id.ascending_order : R.id.descending_order);

        groupOrder.addOnButtonCheckedListener((group, checkedId, isChecked) -> {
            if (isChecked) checkedOrderButton = checkedId == R.id.ascending_order ? 0 : 1;
        });

        cancel.setOnClickListener(view -> dismiss());
        apply.setOnClickListener(view -> {

            editor.putInt("SORT", checkedSortButton).apply();
            editor.putInt("ORDER", checkedOrderButton).apply();
            dismiss();
            new Handler().post(this::sort);
        });

    }

    private boolean isNormalScreen(Context context) {
        DisplayMetrics dm = new DisplayMetrics();
        instance.getWindowManager().getDefaultDisplay().getMetrics(dm);
        //double x = Math.pow(mWidthPixels/dm.xdpi,2);
        //double y = Math.pow(mHeightPixels/dm.ydpi,2);
        //double screenInches = Math.sqrt(x+y);
        return context.getResources().getConfiguration()
                .isLayoutSizeAtLeast(Configuration.SCREENLAYOUT_SIZE_NORMAL);
    }

    @SuppressLint("NotifyDataSetChanged")
    public void sort() {
        if (jFileList != null && !jFileList.isEmpty()) compare(jFileList, sort, order == 1);
        jFileAdapter.notifyDataSetChanged();
    }

    public static void sort(ArrayList<JFile> jFileList) {
        if (jFileList != null && !jFileList.isEmpty()) compare(jFileList, sort, order == 1);
    }

    @SuppressLint("NotifyDataSetChanged")
    public void sortAndNotify(JFileAdapter jFileAdapter) {
        if (jFileAdapter.getItemCount() > 0) compare(jFileAdapter.jFileList, sort, order == 1);
        jFileAdapter.notifyDataSetChanged();
    }

    @SuppressLint("NotifyDataSetChanged")
    public void compareAndNotify(JFileAdapter jFileAdapter, int sort, boolean reverse) {
        compare(jFileAdapter.jFileList, sort, reverse);
        jFileAdapter.notifyDataSetChanged();
    }

    // TODO comparings

    public static void compare(ArrayList<JFile> jFiles, int sort, boolean reverse) {
        switch (sort) {
            case 0:
                if (reverse) jFiles.sort(DialogSort::compareNameReverse);
                else jFiles.sort(DialogSort::compareName);
                break;
            case 1:
                if (reverse) jFiles.sort(DialogSort::compareSizeReverse);
                else jFiles.sort(DialogSort::compareSize);
                break;
            case 2:
                if (reverse) jFiles.sort(DialogSort::compareDateReverse);
                else jFiles.sort(DialogSort::compareDate);
                break;
            case 3:
                if (reverse) jFiles.sort(DialogSort::compareTypeReverse);
                else jFiles.sort(DialogSort::compareType);
                break;
        }
    }

    public static void compareRecentFile(ArrayList<JFile> jFiles) {
        jFiles.sort(DialogSort::compareRecentFile);
    }

    public static int compareRecentFile(JFile file1, JFile file2) {

        return Long.compare(file2.lastModified(), file1.lastModified());
    }

    public static int compareName(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        return compare(file1.getNameTLC(), file2.getNameTLC());
    }

    public static int compareNameReverse(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        return compare(file2.getNameTLC(), file1.getNameTLC());
    }

    public static int compareSize(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        int sizeOrder = Long.compare(file1.getSize(), file2.getSize());
        if (sizeOrder != 0) return sizeOrder;
        else return compare(file1.getNameTLC(), file2.getNameTLC());
    }

    public static int compareSizeReverse(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        int sizeOrder = Long.compare(file2.getSize(), file1.getSize());
        if (sizeOrder != 0) return sizeOrder;
        else return compare(file1.getNameTLC(), file2.getNameTLC());
    }

    public static int compareDate(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        return Long.compare(file1.lastModified(), file2.lastModified());
    }

    public static int compareDateReverse(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        return Long.compare(file2.lastModified(), file1.lastModified());
    }

    public static int compareType(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        int extOrder = file1.getExtension().compareTo(file2.getExtension());
        if (extOrder != 0) return extOrder;
        else return compare(file1.getNameTLC(), file2.getNameTLC());
    }

    public static int compareTypeReverse(JFile file1, JFile file2) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        int extOrder = file2.getExtension().compareTo(file1.getExtension());
        if (extOrder != 0) return extOrder;
        else return compare(file1.getNameTLC(), file2.getNameTLC());
    }

    public static int compare(JFile file1, JFile file2, int sort, boolean reverse) {

        if (file1.isDirectory() && !file2.isDirectory()) {
            return -1;
        } else if (!file1.isDirectory() && file2.isDirectory()) {
            return 1;
        }

        if (sort == 0) {

            // sort by name
            if (reverse) return compare(file2.getNameTLC(),
                    file1.getNameTLC());
            else return compare(file1.getNameTLC(),
                    file2.getNameTLC());
        } else if (sort == 1) {

            // sort by size
            int sizeOrder;
            if (reverse) sizeOrder = Long.compare(file2.getSize(), file1.getSize());
            else sizeOrder = Long.compare(file1.getSize(), file2.getSize());
            if (sizeOrder != 0) return sizeOrder;
            else return compare(file1.getNameTLC(),
                    file2.getNameTLC());
        } else if (sort == 2) {

            // sort by last modified
            if (reverse) return Long.compare(file2.lastModified(), file1.lastModified());
            else return Long.compare(file1.lastModified(), file2.lastModified());
        } else if (sort == 3) {

            // sort by type
            int extOrder;
            if (reverse) extOrder = file2.getExtension().compareTo(file1.getExtension());
            else extOrder = file1.getExtension().compareTo(file2.getExtension());
            if (extOrder != 0) return extOrder;
            else return compare(file1.getNameTLC(),
                    file2.getNameTLC());
        }

        return 0;
    }

    // http://www.davekoelle.com/files/AlphanumComparator.java
    private static boolean isDigit(char ch) {
        return ((ch >= 48) && (ch <= 57));
    }

    /**
     * Length of string is passed in for improved efficiency (only need to calculate it once)
     **/
    private static String getChunk(String s, int sLength, int marker) {
        StringBuilder chunk = new StringBuilder();
        char c = s.charAt(marker);
        chunk.append(c);
        marker++;
        if (isDigit(c)) {
            while (marker < sLength) {
                c = s.charAt(marker);
                if (!isDigit(c))
                    break;
                chunk.append(c);
                marker++;
            }
        } else {
            while (marker < sLength) {
                c = s.charAt(marker);
                if (isDigit(c))
                    break;
                chunk.append(c);
                marker++;
            }
        }
        return chunk.toString();
    }

    public static int compare(String s1, String s2) {
        if ((s1 == null) || (s2 == null)) {
            return 0;
        }

        int thisMarker = 0;
        int thatMarker = 0;
        int s1Length = s1.length();
        int s2Length = s2.length();

        while (thisMarker < s1Length && thatMarker < s2Length) {
            String thisChunk = getChunk(s1, s1Length, thisMarker);
            thisMarker += thisChunk.length();

            String thatChunk = getChunk(s2, s2Length, thatMarker);
            thatMarker += thatChunk.length();

            // If both chunks contain numeric characters, sort them numerically
            int result;
            if (isDigit(thisChunk.charAt(0)) && isDigit(thatChunk.charAt(0))) {
                // Simple chunk comparison by length.
                int thisChunkLength = thisChunk.length();
                result = thisChunkLength - thatChunk.length();
                // If equal, the first different number counts
                if (result == 0) {
                    for (int i = 0; i < thisChunkLength; i++) {
                        result = thisChunk.charAt(i) - thatChunk.charAt(i);
                        if (result != 0) {
                            return result;
                        }
                    }
                }
            } else {
                result = collator.compare(thisChunk, thatChunk);
            }

            if (thisChunk.startsWith(".") && !thatChunk.startsWith(".")) result = 1;
            else if (!thisChunk.startsWith(".") && thatChunk.startsWith(".")) result = -1;

            if (result != 0)
                return result;
        }

        return s1Length - s2Length;
    }

}

File: app/src/main/java/com/example/files/actions/DialogSort.kt
package com.example.files.actions

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.icu.text.Collator
import android.os.Handler
import android.os.Looper
import android.util.DisplayMetrics
import android.view.ViewGroup
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.files.JFileAdapter
import com.example.files.MainActivity.editor
import com.example.files.MainActivity.instance
import com.example.files.R
import com.example.files.Statics.order
import com.example.files.Statics.sort
import com.example.files.models.JFile
import com.google.android.material.bottomsheet.BottomSheetDialog
import java.util.*

class DialogSort : BottomSheetDialog {

    private val activity: Activity
    private var jFileList: ArrayList<JFile>? = null
    private var jFileAdapter: JFileAdapter? = null
    private val isNormalSize: Boolean

    constructor() : super(instance) {
        this.activity = instance
        this.isNormalSize = isNormalScreen(activity)
    }

    constructor(jFileAdapter: JFileAdapter) : super(instance) {
        this.activity = instance
        this.jFileAdapter = jFileAdapter
        this.jFileList = jFileAdapter.jFileList
        this.isNormalSize = isNormalScreen(activity)
        createDialogSort()
    }

    private fun createDialogSort() {
        val composeView = ComposeView(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            setContent {
                MaterialTheme {
                    SortDialogContent(
                        initialSort = sort,
                        initialOrder = order,
                        onApply = { selectedSort, selectedOrder ->
                            editor.putInt("SORT", selectedSort).apply()
                            editor.putInt("ORDER", selectedOrder).apply()
                            dismiss()
                            Handler(Looper.getMainLooper()).post { sort() }
                        },
                        onCancel = { dismiss() }
                    )
                }
            }
        }
        setContentView(composeView)
    }

    private fun isNormalScreen(context: Context): Boolean {
        return context.resources.configuration.isLayoutSizeAtLeast(Configuration.SCREENLAYOUT_SIZE_NORMAL)
    }

    @SuppressLint("NotifyDataSetChanged")
    fun sort() {
        if (!jFileList.isNullOrEmpty()) {
            compare(jFileList!!, sort, order == 1)
        }
        jFileAdapter?.notifyDataSetChanged()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun sortAndNotify(jFileAdapter: JFileAdapter) {
        if (jFileAdapter.itemCount > 0) {
            compare(jFileAdapter.jFileList, sort, order == 1)
        }
        jFileAdapter.notifyDataSetChanged()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun compareAndNotify(jFileAdapter: JFileAdapter, sort: Int, reverse: Boolean) {
        compare(jFileAdapter.jFileList, sort, reverse)
        jFileAdapter.notifyDataSetChanged()
    }

    companion object {
        private val collator = Collator.getInstance(Locale.getDefault())

        @JvmStatic
        fun sort(jFileList: ArrayList<JFile>?) {
            if (!jFileList.isNullOrEmpty()) {
                compare(jFileList, sort, order == 1)
            }
        }

        @JvmStatic
        fun compare(jFiles: ArrayList<JFile>, sortType: Int, reverse: Boolean) {
            when (sortType) {
                0 -> if (reverse) jFiles.sortWith(Comparator { f1, f2 -> compareNameReverse(f1, f2) }) else jFiles.sortWith(Comparator { f1, f2 -> compareName(f1, f2) })
                1 -> if (reverse) jFiles.sortWith(Comparator { f1, f2 -> compareSizeReverse(f1, f2) }) else jFiles.sortWith(Comparator { f1, f2 -> compareSize(f1, f2) })
                2 -> if (reverse) jFiles.sortWith(Comparator { f1, f2 -> compareDateReverse(f1, f2) }) else jFiles.sortWith(Comparator { f1, f2 -> compareDate(f1, f2) })
                3 -> if (reverse) jFiles.sortWith(Comparator { f1, f2 -> compareTypeReverse(f1, f2) }) else jFiles.sortWith(Comparator { f1, f2 -> compareType(f1, f2) })
            }
        }

        @JvmStatic
        fun compareRecentFile(jFiles: ArrayList<JFile>) {
            jFiles.sortWith(Comparator { f1, f2 -> compareRecentFile(f1, f2) })
        }

        @JvmStatic
        fun compareRecentFile(file1: JFile, file2: JFile): Int {
            return java.lang.Long.compare(file2.lastModified(), file1.lastModified())
        }

        @JvmStatic
        fun compareName(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            return compare(file1.nameTLC, file2.nameTLC)
        }

        @JvmStatic
        fun compareNameReverse(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            return compare(file2.nameTLC, file1.nameTLC)
        }

        @JvmStatic
        fun compareSize(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            val sizeOrder = java.lang.Long.compare(file1.size, file2.size)
            return if (sizeOrder != 0) sizeOrder else compare(file1.nameTLC, file2.nameTLC)
        }

        @JvmStatic
        fun compareSizeReverse(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            val sizeOrder = java.lang.Long.compare(file2.size, file1.size)
            return if (sizeOrder != 0) sizeOrder else compare(file1.nameTLC, file2.nameTLC)
        }

        @JvmStatic
        fun compareDate(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            return java.lang.Long.compare(file1.lastModified(), file2.lastModified())
        }

        @JvmStatic
        fun compareDateReverse(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            return java.lang.Long.compare(file2.lastModified(), file1.lastModified())
        }

        @JvmStatic
        fun compareType(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            val extOrder = file1.extension.compareTo(file2.extension)
            return if (extOrder != 0) extOrder else compare(file1.nameTLC, file2.nameTLC)
        }

        @JvmStatic
        fun compareTypeReverse(file1: JFile, file2: JFile): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1
            val extOrder = file2.extension.compareTo(file1.extension)
            return if (extOrder != 0) extOrder else compare(file1.nameTLC, file2.nameTLC)
        }

        @JvmStatic
        fun compare(file1: JFile, file2: JFile, sort: Int, reverse: Boolean): Int {
            if (file1.isDirectory && !file2.isDirectory) return -1
            else if (!file1.isDirectory && file2.isDirectory) return 1

            return when (sort) {
                0 -> if (reverse) compare(file2.nameTLC, file1.nameTLC) else compare(file1.nameTLC, file2.nameTLC)
                1 -> {
                    val sizeOrder = if (reverse) java.lang.Long.compare(file2.size, file1.size) else java.lang.Long.compare(file1.size, file2.size)
                    if (sizeOrder != 0) sizeOrder else compare(file1.nameTLC, file2.nameTLC)
                }
                2 -> if (reverse) java.lang.Long.compare(file2.lastModified(), file1.lastModified()) else java.lang.Long.compare(file1.lastModified(), file2.lastModified())
                3 -> {
                    val extOrder = if (reverse) file2.extension.compareTo(file1.extension) else file1.extension.compareTo(file2.extension)
                    if (extOrder != 0) extOrder else compare(file1.nameTLC, file2.nameTLC)
                }
                else -> 0
            }
        }

        private fun isDigit(ch: Char): Boolean = ch in '0'..'9'

        private fun getChunk(s: String, sLength: Int, markerArg: Int): String {
            var marker = markerArg
            val chunk = StringBuilder()
            var c = s[marker]
            chunk.append(c)
            marker++
            if (isDigit(c)) {
                while (marker < sLength) {
                    c = s[marker]
                    if (!isDigit(c)) break
                    chunk.append(c)
                    marker++
                }
            } else {
                while (marker < sLength) {
                    c = s[marker]
                    if (isDigit(c)) break
                    chunk.append(c)
                    marker++
                }
            }
            return chunk.toString()
        }

        @JvmStatic
        fun compare(s1: String?, s2: String?): Int {
            if (s1 == null || s2 == null) return 0
            var thisMarker = 0
            var thatMarker = 0
            val s1Length = s1.length
            val s2Length = s2.length

            while (thisMarker < s1Length && thatMarker < s2Length) {
                val thisChunk = getChunk(s1, s1Length, thisMarker)
                thisMarker += thisChunk.length

                val thatChunk = getChunk(s2, s2Length, thatMarker)
                thatMarker += thatChunk.length

                var result: Int
                if (isDigit(thisChunk[0]) && isDigit(thatChunk[0])) {
                    val thisChunkLength = thisChunk.length
                    result = thisChunkLength - thatChunk.length
                    if (result == 0) {
                        for (i in 0 until thisChunkLength) {
                            result = thisChunk[i] - thatChunk[i]
                            if (result != 0) return result
                        }
                    }
                } else {
                    result = collator.compare(thisChunk, thatChunk)
                }

                if (thisChunk.startsWith(".") && !thatChunk.startsWith(".")) result = 1
                else if (!thisChunk.startsWith(".") && thatChunk.startsWith(".")) result = -1

                if (result != 0) return result
            }
            return s1Length - s2Length
        }
    }
}

@Composable
fun SortDialogContent(
    initialSort: Int,
    initialOrder: Int,
    onApply: (Int, Int) -> Unit,
    onCancel: () -> Unit
) {
    var selectedSort by remember { mutableIntStateOf(initialSort) }
    var selectedOrder by remember { mutableIntStateOf(initialOrder) }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp)
    ) {
        Text(
            text = stringResource(id = R.string.sort_by),
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 5.dp, vertical = 0.dp)
                .padding(bottom = 20.dp)
        )

        SingleChoiceSegmentedButtonRow(
            modifier = Modifier.fillMaxWidth()
        ) {
            val sortOptions = listOf(
                0 to stringResource(R.string.name),
                1 to stringResource(R.string.size),
                2 to stringResource(R.string.sort_date),
                3 to stringResource(R.string.sort_type)
            )

            sortOptions.forEachIndexed { index, (id, label) ->
                SegmentedButton(
                    shape = SegmentedButtonDefaults.itemShape(index = index, count = sortOptions.size),
                    onClick = { selectedSort = id },
                    selected = selectedSort == id
                ) {
                    Text(text = label)
                }
            }
        }

        Spacer(modifier = Modifier.height(10.dp))

        Text(
            text = stringResource(R.string.order),
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier
                .fillMaxWidth()
                .padding(start = 8.dp, top = 10.dp, bottom = 5.dp)
        )

        SingleChoiceSegmentedButtonRow(
            modifier = Modifier.fillMaxWidth().padding(top = 10.dp)
        ) {
            val orderOptions = listOf(
                0 to stringResource(R.string.ascending_order),
                1 to stringResource(R.string.descending_order)
            )

            orderOptions.forEachIndexed { index, (id, label) ->
                SegmentedButton(
                    shape = SegmentedButtonDefaults.itemShape(index = index, count = orderOptions.size),
                    onClick = { selectedOrder = id },
                    selected = selectedOrder == id
                ) {
                    Text(text = label)
                }
            }
        }

        Spacer(modifier = Modifier.height(20.dp))

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            TextButton(
                onClick = onCancel,
                modifier = Modifier.weight(1f)
            ) {
                Text(text = stringResource(R.string.cancel), fontSize = 17.sp, fontWeight = FontWeight.Bold)
            }

            HorizontalDivider(
                modifier = Modifier
                    .width(2.dp)
                    .height(24.dp)
                    .padding(horizontal = 5.dp)
            )

            TextButton(
                onClick = { onApply(selectedSort, selectedOrder) },
                modifier = Modifier.weight(1f)
            ) {
                Text(text = stringResource(R.string.apply), fontSize = 17.sp, fontWeight = FontWeight.Bold)
            }
        }
    }
}
